specs/codecs/dag-cosmos/tendermint_chain.md

```diff
@@ -0,0 +1,355 @@
+
+# Header defines the structure of a Tendermint block header
+type Header struct {
+	# basic block info
+	Version Version
+	ChainID String
+	Height  Int
+	Time    Time
+
+	# prev block info
+	# the first block has an empty BlockID struct as LastBlockID
+	LastBlockID BlockID
+
+	# hashes of block data
+	LastCommitHash CommitTreeCID # MerkleRoot of the current Signatures
+	DataHash       TxTreeCID # MerkleRoot of the current Txs
+
+	# hashes from the app output from the prev block
+	ValidatorsHash     ValidatorTreeCID # MerkleRoot of the current validator set
+	NextValidatorsHash ValidatorTreeCID # MerkleRoot of the next validator set
+	ConsensusHash      HashedParamsCID # SHA256 hash of HashedParams
+	AppHash            AppStateTreeCID # State Root from the state machine
+
+	# root hash of all results from the txs from the previous block
+	LastResultsHash ResultTreeCID # MerkleRoot of the last blocks Result set
+
+	# consensus info
+	EvidenceHash    EvidenceTreeCID # MerkleRoot of the current Evidence set
+	ProposerAddress Address
+}
+```
+
+## Tendermint Params
+```ipldsch
+# HashedParamsCID is a CID link to the HashedParams for this Header
+# This CID is composed of the SHA_256 multihash of the linked protobuf encoded HashedParams struct and the Params codec (tbd)
+type HashParamsCID &HashedParams
+
+# HashedParams is a subset of ConsensusParams that is included in the consensus encoding
+# It is hashed into the Header.ConsensusHash
+type HashedParams struct {
+	BlockMaxBytes Int
+	BlockMaxGas   Int
+}
+```
+
+## Tendermint Result
+```ipldsch
+# ResultTreeCID is a CID link to the root node of a Result merkle tree
+# This CID is composed of the SHA_256 multihash of the root node in a Result merkle tree and the ResultTree codec (tbd)
+# Result merkle tree is a Merkle tree built from ResponseDeliverTx responses (Log, Info, Codespace and Events fields are ignored)
+type ResultTreeCID &MerkleTreeNode
+
+# ResponseDeliverTx are the consensus fields of an ABCI ResponseDeliverTx, that are included in the ResultTree
+type ResponseDeliverTx struct {
+    Code      Uint
+    Data      Bytes
+    GasWanted Int
+    GasUsed   Int
+}
+```
+
+## Tendermint Transaction
+```ipldsch
+# TxTreeCID is a CID link to the root node of a Tx merkle tree
+# This CID is composed of the SHA_256 multihash of the root node in the Tx merkle tree and the TxTree codec (tbd)
+# Tx merkle tree is a Merkle tree built from the set of Txs at the given block
+# Note: The transactions are hashed before being included in the Merkle tree, the leaves of the Merkle tree are the hashes, not the transactions themselves.
+type TxTreeCID &MerkleTreeNode
+
+#Txs is a list of Tx.
+type Txs [Tx]
+
+# TxCID is a link to a TX
+# TxCID uses the SHA256 multihash and the Raw codec (0x55)
+type TxCID &Tx
+
+# Tx is an arbitrary byte array.
+# Tx has no types at the Tendermint level, instead it encodes arbitrary length-prefixed data that can be decoded and processed in a state-machine specific manner.
```

Comment:
! this is pretty frustrating .. I guess this is one of the points where you need the adjacent data structures just to piece the data together again into a consistent graph

Reply:
Yes I am really struggling with how to represent this more rich type information. These txs map to a single cosmos state machine, the state machine which can make sense of these arbitrary bytes. So we could create a "namespacing" system for different Tendermint/Cosmos blockchains on IPLD, and use that to map these txs to the state machine that processes them. I'm not sure if that would involve registering a multicodec type for every Cosmos blockchain, or if we could take an approach similar to the one proposed here for supporting arbitrary protobuf objects where we prepend these bytes with a hash reference to an object type that provides that namespacing.

Reply:
Hmm actually I retract that statement, as I don't think we achieve any better typing than we already have. The `Header` IPLD has `ChainID` and `Version` fields, and these two fields specify the type of state machine and its specific version that is capable of making sense of the tx bytes.

Reply:
Aaand I retract my retraction, if we assume CosmosSDK standards are followed- which we already do elsewhere- then these tx bytes are always serialized protobuf messages and we can provide richer typing using the same `TypedProtobuf` scheme we propose here for typing arbitrary values stored in IAVL or SMT state. Updated the spec to reflect this, as well as cleaning things up some more.